assets/js/base/context/cart-checkout/shipping/index.js

```diff
@@ -154,36 +121,71 @@ export const ShippingDataProvider = ( { children } ) => {
 	// emit events
 	// @todo add emitters for shipping rate selection.
 	useEffect( () => {
-		if ( ! shippingOptionsLoading && currentErrorStatus.hasError ) {
+		if ( ! shippingRatesLoading && currentErrorStatus.hasError ) {
 			emitEvent(
 				currentObservers.current,
-				EMIT_TYPES.SHIPPING_RATES_SUCCESS,
+				EMIT_TYPES.SHIPPING_RATES_FAIL,
 				shippingErrorStatus
 			);
-		} else if ( ! shippingOptionsLoading && shippingOptions ) {
+		} else if ( ! shippingRatesLoading && shippingRates ) {
 			emitEvent(
 				currentObservers.current,
 				EMIT_TYPES.SHIPPING_RATES_SUCCESS,
-				shippingOptions
+				shippingRates
 			);
 		}
 	}, [
-		shippingOptions,
-		shippingOptionsLoading,
+		shippingRates,
+		shippingRatesLoading,
 		currentErrorStatus,
 		shippingErrorStatus,
 	] );
 
+	//emit shipping rate selection events
+	useEffect( () => {
+		if ( ! isSelectingRate && currentErrorStatus.hasError ) {
+			emitEvent(
+				currentObservers.current,
+				EMIT_TYPES.SHIPPING_RATE_SELECT_FAIL,
+				shippingErrorStatus
+			);
+		} else if ( ! isSelectingRate && selectedRates ) {
+			emitEvent(
+				currentObservers.current,
+				EMIT_TYPES.SHIPPING_RATE_SELECT_SUCCESS,
+				selectedRates
+			);
+		}
+	}, [
+		selectedRates,
+		isSelectingRate,
+		currentErrorStatus,
+		shippingErrorStatus,
+	] );
+
+	// dispatch checkout error if there's a shipping error.
+	useEffect( () => {
+		if ( currentErrorStatus.hasError ) {
+			dispatchActions.setHasError();
+		} else {
+			dispatchActions.clearError();
```

Comment:
Just a comment on these events; I checked and we're not storing errors here, just whether or not there is an error. Whats the point in having 2 sets of ACTIONS and EVENTS to `setHasError` and `clearError` (which has a differently named action set_no_error) when we could just have  function accepting a boolean? `dispatchActions.hasError( true | false );` Would make this easier to understand.

Reply:
I'll take care of in a followup (see #2081)